File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/ports/PaymentProviderResult.java
package it.giannibombelli.wsc2026.payment.domain.ports;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;

import java.util.UUID;

public sealed interface PaymentProviderResult {
    /**
     * {@code providerCompletedAt} è il dato fornito dal provider che determina la finestra delle 48h
     * nell'aggregato Payment: non è il wall-clock del comando AcceptTransaction.
     */
    record Success(UUID transactionId, Timestamp providerCompletedAt) implements PaymentProviderResult {
        public Success {
            Require.requireArgument(transactionId, "transactionId");
            Require.requireArgument(providerCompletedAt, "providerCompletedAt");
        }
    }

    record Failure(Description reason) implements PaymentProviderResult {
        public Failure {
            Require.requireArgument(reason, "reason");
        }
    }
}

File: typescript/src/payment/domain/ports/paymentProviderResult.ts
// Risultato restituito da un PaymentProvider.

import { Description } from '@/common/domain/primitive/description.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type PaymentProviderResult =
  | { readonly kind: 'success'; readonly transactionId: Uuid; readonly providerCompletedAt: Timestamp }
  | { readonly kind: 'failure'; readonly reason: Description };

export function paymentProviderSuccess(transactionId: Uuid, providerCompletedAt: Timestamp): PaymentProviderResult {
  requireArgument(transactionId, 'transactionId');
  requireArgument(providerCompletedAt, 'providerCompletedAt');
  return { kind: 'success', transactionId, providerCompletedAt };
}

export function paymentProviderFailure(reason: Description): PaymentProviderResult {
  requireArgument(reason, 'reason');
  return { kind: 'failure', reason };
}
